### PR TITLE
Add compile-time support for `dict.keys`

### PR DESCRIPTION
**Pull request by kmr-srbh**
### Working

```python
from lpython import i32


print({1: "a"}.keys())
print({"a": 1, "b": 2, "c": 3}.keys())
print({(1, 2): "a", (3, 4): "b", (5, 6): "c"}.keys())

k_1: list[i32] = {
    1: [1, 2, 3],
    2: [4, 5, 6],
    3: [7, 8, 9],
}.keys()
print(k_1)

k_2: list[tuple[i32, i32]] = {
    (1, 2): "a",
    (3, 4): "b",
    (5, 6): "c",
}.keys()
print(k_2)
```

#### CPython
```console
(base) saurabh-kumar@Awadh:~/Projects/System/lpython$ python ./examples/example.py
dict_keys([1])
dict_keys(['a', 'b', 'c'])
dict_keys([(1, 2), (3, 4), (5, 6)])
dict_keys([1, 2, 3])
dict_keys([(1, 2), (3, 4), (5, 6)])
```

#### LPython
```console
(base) saurabh-kumar@Awadh:~/Projects/System/lpython$ ./src/bin/lpython ./examples/example.py
[1]
['a', 'b', 'c']
[(1, 2), (3, 4), (5, 6)]
[1, 2, 3]
[(1, 2), (3, 4), (5, 6)]
```

### CODE REVIEW

**Comment by kmr-srbh** [comment]
@Shaikh-Ubaid , all the tests pass on my local branch. Why does it fail here?

**Comment by kmr-srbh** [comment]
Got the issue @Shaikh-Ubaid! You are absolutely right about the interoperability. The issue is related to the way I am testing the method. The output of `dict.keys` is an object view in CPython, but we do not support that yet. I am pushing new tests that check the length instead.

File: src/lpython/semantics/python_attribute_eval.h [hunk 1]
```diff
@@ -74,13 +74,13 @@ struct AttributeHandler {
...
-                            + std::string(v->m_name) + "` is not allowed", loc);
-            }
-        }
+        // if (modify_attr_set.find(key) != modify_attr_set.end()) {
+        //     ASR::Variable_t* v = ASRUtils::EXPR2VAR(e);
+        //     if (v->m_intent == ASRUtils::intent_in) {
+        //         throw SemanticError("Modifying input function parameter `"
+        //                     + std::string(v->m_name) + "` is not allowed", loc);
+        //     }
+        // }
```
Comment (Shaikh-Ubaid) [inline src/lpython/semantics/python_attribute_eval.h:83]: This should not be commented.

**Comment by kmr-srbh** [comment]
> @kmr-srbh do you mind posting the output for `cpython` for the example shared in the PR description?

I am posting the output @Shaikh-Ubaid. :+1:

File: integration_tests/test_dict_keys_values.py [hunk 2]
```diff
@@ -51,4 +51,27 @@ def test_dict_keys_values():
                 assert v2_copy[j] == d2[str(i)]
         assert key_count == 1
 
+
+    # dict.keys on dict constant
+    assert len({1: "a"}.keys()) == 1
```
Comment (Shaikh-Ubaid) [inline integration_tests/test_dict_keys_values.py:56]: Can you update the tests according to https://github.com/lcompilers/lpython/pull/2661#discussion_r1580436211?

File: tests/reference/asr-func_04-eef2656.stdout [hunk 3]
```diff
@@ -0,0 +1,82 @@
+(TranslationUnit
```
Comment (Shaikh-Ubaid) [inline tests/reference/asr-func_04-eef2656.stdout:1]: This file should be deleted.

**Review by Shaikh-Ubaid — COMMENTED** [review]
Shared two comments. Rest looks good to me. Thanks for this! Great work!

File: integration_tests/test_dict_keys_values.py (hunk 2 above)
Comment (kmr-srbh) [reply integration_tests/test_dict_keys_values.py:56]: I had initially tried this @Shaikh-Ubaid , but it throws the error as stated by you in #2667. I am looking into it.

File: src/lpython/semantics/python_ast_to_asr.cpp [hunk 4]
```diff
@@ -7597,6 +7597,20 @@ we will have to use something else.
                     }
                 }
             }
+        } else if (AST::is_a<AST::Dict_t>(*at->m_value)) {
+            AST::Dict_t* cdict = AST::down_cast<AST::Dict_t>(at->m_value);
+            visit_Dict(*cdict);
+            if (!tmp) {
```
Comment (Shaikh-Ubaid) [inline src/lpython/semantics/python_ast_to_asr.cpp:7603]:
```suggestion
            if (tmp == nullptr) {
```

File: src/lpython/semantics/python_ast_to_asr.cpp [hunk 5]
```diff
@@ -7597,6 +7597,20 @@ we will have to use something else.
...
+                throw SemanticError("cannot call " + std::string(at->m_attr) + " on an empty dict" , loc);
+            }
+            ASR::expr_t* dict_expr = ASR::down_cast<ASR::expr_t>(tmp);
+            Vec<ASR::expr_t*> eles;
+            eles.reserve(al, args.size());
+            for (size_t i=0; i<args.size(); i++) {
```
Comment (Shaikh-Ubaid) [inline src/lpython/semantics/python_ast_to_asr.cpp:7609]:
```suggestion
            for (size_t i = 0; i < args.size(); i++) {
```

**Comment by Shaikh-Ubaid** [comment]
Overall, it seems fine to me. Let's wait on this until https://github.com/lcompilers/lpython/pull/2661#issuecomment-2080395795 is clear.

**Review by Shaikh-Ubaid — APPROVED** [review]
It looks good to me. Thanks for this!

**Pull request merged**